File: runme/mcgen/local/datan/exasrc/e8lsq.f
      PROGRAM E8LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXN=8,MAXNR=2,MAXM=4)
      COMMON /GRVARS/ X(MAXNR),Y(MAXN),CY(MAXN,MAXN),YY(MAXN),
     +GY(MAXN,MAXN),CX(MAXNR,MAXNR),CYY(MAXN,MAXN),
     +F(MAXN+MAXNR,MAXN+MAXNR),E(MAXM,MAXN+MAXNR),
     +A2(MAXN+MAXNR,MAXN+MAXNR),LIST(2),DXPLUS(2),DXMINS(2),
     +XPL(2),YPL(2)
      DIMENSION T(4),S(4),DT(4),DS(4),RHO(4)
      CHARACTER*75 STRING
      CHARACTER*1 ANSWER,YES1,YES2
      DATA YES1/'Y'/
      DATA YES2/'y'/
      DATA T /0.2D0,0.2D0,0.8D0,0.85D0/
      DATA S /0.15D0,0.7D0,1.1D0,0.8D0/
      DATA DT/0.4D0,0.4D0,0.4D0,0.2D0/
      DATA DS/0.2D0,0.4D0,0.4D0,0.4D0/
      DATA RHO/0.D0,0.D0,0.5D0,0.D0/
C identify program to user
      WRITE(*,*)' Program E8LSQ demonstrates'
      WRITE(*,*)' use of LSQASG and LSQCOG'
      WRITE(*,*)' '
C write table of data
      WRITE(*,*)'    T         S         DT        DS        RHO'
      DO 5 I=1,4
        WRITE(*,'(5F10.5)')T(I),S(I),DT(I),DS(I),RHO(I)
5     CONTINUE
      WRITE(*,'(/)')
C set up data for input to LSQGEN
      N=8
      M=4
      NR=2
      CALL MTXUNT(CY,8)
      DO 10 I=1,4
        Y((I-1)*2+1)=T(I)
        Y((I-1)*2+2)=S(I)
        CY((I-1)*2+1,(I-1)*2+1)=DT(I)**2
        CY((I-1)*2+2,(I-1)*2+2)=DS(I)**2
        CY((I-1)*2+1,(I-1)*2+2)=RHO(I)*DS(I)*DT(I)
        CY((I-1)*2+2,(I-1)*2+1)=RHO(I)*DS(I)*DT(I)
10    CONTINUE
C save input data
      CALL MTXCPV(Y,YY,N)
      CALL MTXTRA(CY,CYY,N,N)
C determine first approximation
      NRED=2
      LIST(1)=1
      LIST(2)=1
      X(2)=(S(4)-S(1))/(T(4)-T(1))
      X(1)=S(1)-X(2)*T(1)
C perform fit with LSQGEN
      WRITE(*,*)' performing fit with LSQGEN'
      WRITE(*,'(3(A,I2),A,2I2)')' N = ',N,', NR = ',NR,', NRED = ',
     +NRED,', LIST = ',LIST
      WRITE(*,'(A,2F10.5)')' first approx.: X = ',X
      NSTEP=100
      CALL LSQGEN(Y,CY,GY,F,E,M,N,NR,NRED,LIST,X,CX,R,A2,NSTEP)
      WRITE(*,'(A,F6.2,A,I3,/,A,(2F10.5))')' result of fit: R = ',R,
     +', NSTEP =',NSTEP,' X =',X
      WRITE(*,*)' covariance matrix CX = '
      CALL MTXWRT(CX,NRED,NRED)
      WRITE(*,'(/)')
C store solution, errors and correlation for later drawing
      X1=X(1)
      X2=X(2)
      DX1=SQRT(CX(1,1))
      DX2=SQRT(CX(2,2))
      RH=CX(1,2)/(DX1*DX2)
C restore input data and determine asymmetric errors
      CALL MTXCPV(YY,Y,N)
      CALL MTXTRA(CYY,CY,N,N)
      NSTEP=100
      W=0.D0
      CALL LSQASG(Y,CY,GY,F,E,M,N,NR,NRED,LIST,X,CX,R,W,DXPLUS,
     +DXMINS,A2,NSTEP)
      WRITE(*,*)' asymmetric errors:'
      WRITE(*,'(A,2F10.5)')' DXPLUS = ',DXPLUS
      WRITE(*,'(A,2F10.5)')' DXMINS = ',DXMINS
      WRITE(*,*)' press Y to see plot of confidence region'
      WRITE(*,*)' press N to stop  '
      READ(*,'(A)')ANSWER
      IF(ANSWER.EQ.YES1 .OR. ANSWER.EQ.YES2) THEN
C plot confidence region
C ask for number of workstation
        CALL GRNBWS()
        WRITE(*,*)' Please, enter number of workstation:'
        WRITE(*,*)' >'
        READ(*,*)NWS
        CALL GROPEN
        CALL GROPWS(NWS)
        XMIN=-1.5D0
        YMIN=0.D0
        XMAX=1.5D0
        YMAX=3.D0
        CALL GRWNCC(XMIN,XMAX,YMIN,YMAX)
        CALL GRVWWC(0.2D0,0.9D0,0.2D0,0.9D0)
        CALL GRWNWC(-.414D0,1.D0,0.D0,1.D0)
        CALL GRFRAM
        CALL GRBOUN
        STRING='x_1'
        CALL GRSCLX(STRING,75)
        STRING='x_2'
        CALL GRSCLY(STRING,75)
C draw solution with symmetric errors and covariance ellipse
        CALL GRSTCL(2)
        CALL GRDATP(1,.5D0,X1,X2,DX1,DX2,RH)
C draw asymmetric errors
        CALL GRSTCL(3)
        DO 20 I=1,2
          IF(I.EQ.1) THEN
            XPL(1)=X(1)-DXMINS(1)
          ELSE
            XPL(1)=X(1)+DXPLUS(1)
          END IF
          XPL(2)=XPL(1)
          YPL(1)=YMIN
          YPL(2)=YMAX
          CALL GRPLIN(2,XPL,YPL)
20      CONTINUE
        DO 30 I=1,2
          IF(I.EQ.1) THEN
            YPL(1)=X(2)-DXMINS(2)
          ELSE
            YPL(1)=X(2)+DXPLUS(2)
          END IF
          YPL(2)=YPL(1)
          XPL(1)=XMIN
          XPL(2)=XMAX
          CALL GRPLIN(2,XPL,YPL)
30      CONTINUE
        FCONT=R+1.D0
        NX=30
        NY=30
        DX=(XMAX-XMIN)/NX
        DY=(YMAX-YMIN)/NY
        CALL GRSTCL(4)
C retore input data
        CALL MTXCPV(YY,Y,N)
        CALL MTXTRA(CYY,CY,N,N)
C draw confidence region
        CALL LSQCOG(1,2,XMIN,YMIN,DX,DY,NX,NY,FCONT,
     +  Y,CY,GY,F,E,M,N,NR,X,A2)
        CALL GRCLSE
      END IF
      END

File: runme/mcgen/local/datan/exasrc/e8lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Common Block Declarations */

struct grvars {
    double x[2], y[8], cy[64] /* was [8][8] */, yy[8], gy[64] /* was [8][8] */,
           cx[4] /* was [2][2] */, cyy[64] /* was [8][8] */,
           f[100] /* was [10][10] */, e[40] /* was [4][10] */,
           a2[100] /* was [10][10] */;
    integer list[2];
    double dxplus[2], dxmins[2], xpl[2], ypl[2];
} LIBDATA grvars;

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__8 = 8, c__75 = 75;
static double c_dp2 = .2, c_dp9 = .9, c_dmp414 = -.414, c_d1 = 1., c_d0 = 0., 
              c_dp5 = .5;

main(void)
{
    /* Initialized data */ /* Local variables */
    static integer i, m, n, nstep, nred, nr, nx, ny, nws;
    static double t[4] = { .2,.2,.8,.85 }, s[4] = { .15,.7,1.1,.8 },
                  dt[4] = { .4,.4,.4,.2 }, ds[4] = { .2,.4,.4,.4 },
                  rho[4] = { 0.,0.,.5,0. }, xmin, ymin, xmax, ymax,
                  r, w, fcont, x1, x2, rh, dx, dy, dx1, dx2;
    static char yes1[1] = "Y", yes2[1] = "y", string[75], answer[1];

    /* System generated locals */
    double d__1;

/* identify program to user */
    printf("%s\n", "Program E8LSQ demonstrates");
    printf("%s\n\n", "use of LSQASG and LSQCOG");
/* write table of data */
    printf("%s\n", "    T         S         DT        DS        RHO");
    for (i = 1; i <= 4; ++i) {
    printf("%10.5lf%10.5lf%10.5lf%10.5lf%10.5lf\n",
           t[i-1], s[i-1], dt[i-1], ds[i-1], rho[i-1]);
    }
    printf("\n");
/* set up data for input to LSQGEN */
    n = 8;
    m = 4;
    nr = 2;
    mtxunt_(grvars.cy, &c__8);
    for (i = 1; i <= 4; ++i) {
	grvars.y[(i - 1) * 2] = t[i - 1];
	grvars.y[(i - 1 << 1) + 1] = s[i - 1];
/* Computing 2nd power */
	d__1 = dt[i - 1];
	grvars.cy[(i - 1 << 1) + 1 + ((i - 1 << 1) + 1 << 3) - 9] = d__1 * d__1;
/* Computing 2nd power */
	d__1 = ds[i - 1];
	grvars.cy[(i - 1 << 1) + 2 + ((i - 1 << 1) + 2 << 3) - 9] = d__1 * d__1;
	grvars.cy[(i - 1 << 1) + 1 + ((i - 1 << 1) + 2 << 3) - 9] = rho[i - 1]
                 * ds[i - 1] * dt[i - 1];
	grvars.cy[(i - 1 << 1) + 2 + ((i - 1 << 1) + 1 << 3) - 9] = rho[i - 1]
                 * ds[i - 1] * dt[i - 1];
    }
/* save input data */
    mtxcpv_(grvars.y, grvars.yy, &n);
    mtxtra_(grvars.cy, grvars.cyy, &n, &n);
/* determine first approximation */
    nred = 2;
    grvars.list[0] = 1;
    grvars.list[1] = 1;
    grvars.x[1] = (s[3] - s[0]) / (t[3] - t[0]);
    grvars.x[0] = s[0] - grvars.x[1] * t[0];
/* perform fit with LSQGEN */
    printf("%s\n", "performing fit with LSQGEN");
    printf("%s%2li%s%2li%s%2li%s%2li%2li\n", "N = ", n, ", NR = ", nr,
           ", NRED = ", nred, ", LIST = ", grvars.list[0], grvars.list[1]);
    printf("%s%10.5lf%10.5lf\n", "first approx.: X = ",
           grvars.x[0], grvars.x[1]);
    nstep = 100;
    lsqgen_(grvars.y, grvars.cy, grvars.gy, grvars.f, grvars.e, &m, 
	    &n, &nr, &nred, grvars.list, grvars.x, grvars.cx, &r, 
	    grvars.a2, &nstep);
    printf("%s%6.2lf%s%3li\n%s%10.5lf%10.5lf\n", "result of fit: R = ", r,
           ", NSTEP =", nstep, "X =", grvars.x[0], grvars.x[1]);
    printf("%s\n", "covariance matrix CX = ");
    mtxwrt_(grvars.cx, &nred, &nred);
    printf("\n");
/* store solution, errors and correlation for later drawing */
    x1 = grvars.x[0];
    x2 = grvars.x[1];
    dx1 = sqrt(grvars.cx[0]);
    dx2 = sqrt(grvars.cx[3]);
    rh = grvars.cx[2] / (dx1 * dx2);
/* restore input data and determine asymmetric errors */
    mtxcpv_(grvars.yy, grvars.y, &n);
    mtxtra_(grvars.cyy, grvars.cy, &n, &n);
    nstep = 100;
    w = 0.;
    lsqasg_(grvars.y, grvars.cy, grvars.gy, grvars.f, grvars.e, &m, 
	    &n, &nr, &nred, grvars.list, grvars.x, grvars.cx, &r, &w, 
	    grvars.dxplus, grvars.dxmins, grvars.a2, &nstep);
    printf("%s\n", "asymmetric errors:");
    printf("%s", "DXPLUS = ");
    for (i = 1; i <= nred; ++i)
        printf("%10.5lf", grvars.dxplus[i-1]);
    printf("\n");
    printf("%s", "DXMINS = ");
    for (i = 1; i <= nred; ++i)
        printf("%10.5lf", grvars.dxmins[i-1]);
    printf("\n");
    printf("%s\n", "press Y to see plot of confidence region");
    printf("%s\n", "press N to stop  ");
    scanf("%1s", answer);
    if (*answer == yes1[0] || *answer == yes2[0]) {
/* plot confidence region */
/* ask for number of workstation */
        grnbws_();
        printf("%s\n", "Please, enter number of workstation:");
        printf("%s", "> ");
        scanf("%ld", &nws);
        (void)getchar();
	gropen_();
	gropws_(&nws);
	xmin = -1.5;
	ymin = 0.;
	xmax = 1.5;
	ymax = 3.;
	grwncc_(&xmin, &xmax, &ymin, &ymax);
	grvwwc_(&c_dp2, &c_dp9, &c_dp2, &c_dp9);
	grwnwc_(&c_dmp414, &c_d1, &c_d0, &c_d1);
	grfram_();
	grboun_();
        memset(string, ' ', 75);
	strncpy(string, "x_1", 3);
	grsclx_(string, &c__75);
	strncpy(string, "x_2", 3);
	grscly_(string, &c__75);
/* draw solution with symmetric errors and covariance ellipse */
	grstcl_(&c__2);
	grdatp_(&c__1, &c_dp5, &x1, &x2, &dx1, &dx2, &rh);
/* draw asymmetric errors */
	grstcl_(&c__3);
	for (i = 1; i <= 2; ++i) {
	    if (i == 1) {
		grvars.xpl[0] = grvars.x[0] - grvars.dxmins[0];
	    } else {
		grvars.xpl[0] = grvars.x[0] + grvars.dxplus[0];
	    }
	    grvars.xpl[1] = grvars.xpl[0];
	    grvars.ypl[0] = ymin;
	    grvars.ypl[1] = ymax;
	    grplin_(&c__2, grvars.xpl, grvars.ypl);
	}
	for (i = 1; i <= 2; ++i) {
	    if (i == 1) {
		grvars.ypl[0] = grvars.x[1] - grvars.dxmins[1];
	    } else {
		grvars.ypl[0] = grvars.x[1] + grvars.dxplus[1];
	    }
	    grvars.ypl[1] = grvars.ypl[0];
	    grvars.xpl[0] = xmin;
	    grvars.xpl[1] = xmax;
	    grplin_(&c__2, grvars.xpl, grvars.ypl);
	}
	fcont = r + 1.;
	nx = 30;
	ny = 30;
	dx = (xmax - xmin) / nx;
	dy = (ymax - ymin) / ny;
	grstcl_(&c__4);
/* retore input data */
	mtxcpv_(grvars.yy, grvars.y, &n);
	mtxtra_(grvars.cyy, grvars.cy, &n, &n);
/* draw confidence region */
	lsqcog_(&c__1, &c__2, &xmin, &ymin, &dx, &dy, &nx, &ny, &fcont, 
	       grvars.y, grvars.cy, grvars.gy, grvars.f, grvars.e, 
	       &m, &n, &nr, grvars.x, grvars.a2);
	grclse_();
    }
    return 0;
} /* main */

